### DISCUSSION TITLE
z.coerce.bigint() throws error when when using safeParse

### DISCUSSION THREAD

**Discussion by fleck** [colinhacks/zod#1897] [Q&A] [answered]
Given this code:

```ts
const schema = z.coerce.bigint();

const unknownInputFromUser = undefined;

// This throws
schema.safeParse(unknownInputFromUser);
```

schema.safeParse will throw an error.

link to repro: https://codesandbox.io/s/solitary-smoke-39qo2w?file=/src/index.js:49-179

Thank you for zod! Working on a project that uses it currently and it's a delight!

**Comment by JacobWeisenburger** [comment, marked as answer]
https://github.com/colinhacks/zod#coercion-for-primitives
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/BigInt

zod only uses basic coercion. Essentially what is going on is `z.coerce.bigint()` takes what is passed to parse/safeParse and passes it to `BigInt()` and then passes the result to `z.bigint()`.

```ts
BigInt( undefined )
// Uncaught TypeError: Cannot convert undefined to a BigInt
```

So this isn't a bug. It's just a limitation of zod's coercion. 


A workaround is to make your own transform that does the coercion that you want.
```ts
const schema = z.any().transform( value => {
    try { return BigInt( value ) }
    catch ( error ) { return value }
} ).pipe( z.bigint() )

console.log( schema.safeParse( undefined ).success ) // false
```

**Reply by fleck** [reply]
makes sense, thanks for the explanation and workaround!

**Comment by JacobWeisenburger** [comment]
I added an easier way to work with coerce to my [utilz library](https://github.com/JacobWeisenburger/zod_utilz). Perhaps this will help you.

https://github.com/JacobWeisenburger/zod_utilz#coerce

```ts
import { zu } from 'zod_utilz'
const schema = zu.coerce( z.bigint() )
console.log( zu.SPR( schema.safeParse( undefined ) ).error?.issues )
// [
//     {
//         code: 'invalid_type',
//         expected: 'bigint',
//         received: 'undefined',
//         path: [],
//         message: 'Required'
//     }
// ]
```
